org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLSelectionRangesTest.java

```diff
@@ -0,0 +1,506 @@
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(0, 0, 3, 2), r(0, 0, 7, 0)));
+	}
+
+	@Test
+	public void testDoctypeDeclaration4() {
+		String xml = "<!DOCTYPE aaa [\n" + //
+				"  <!ELEMENT aaa (bbb)+>\n" + //
+				"  <!ELEMENT bbb (#PCDATA)>\n" + //
+				"]|>\n" + //
+				"<aaa>\n" + //
+				"  <bbb>Text Content</bbb>\n" + //
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(0, 0, 3, 2), r(0, 0, 7, 0)));
+	}
+
+	@Test
+	public void testDoctypeDeclaration5() {
+		// This test is written so that the behaviour matches that of starting tags
+		String xml = "<!DOCTYPE aaa [\n" + //
+				"  |<!ELEMENT aaa (bbb)+>\n" + //
+				"  <!ELEMENT bbb (#PCDATA)>\n" + //
+				"]>\n" + //
+				"<aaa>\n" + //
+				"  <bbb>Text Content</bbb>\n" + //
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(1, 2, 1, 23), r(0, 15, 3, 0), r(0, 0, 3, 2), r(0, 0, 7, 0)));
+	}
+
+	@Test
+	public void testDoctypeDeclaration6() {
+		String xml = "<!DOCTYPE aaa [\n" + //
+				"  <!ELEMENT aaa (bbb)+>|\n" + //
+				"  <!ELEMENT bbb (#PCDATA)>\n" + //
+				"]>\n" + //
+				"<aaa>\n" + //
+				"  <bbb>Text Content</bbb>\n" + //
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(1, 2, 1, 23), r(0, 15, 3, 0), r(0, 0, 3, 2), r(0, 0, 7, 0)));
+	}
+
+	@Test
+	public void testDoctypeDeclaration7() {
+		String xml = "<!DOCTYPE aaa [\n" + //
+				"  <!ELEMENT aaa (bbb)+>\n" + //
+				"  <!ATTLIST aaa key CDATA \"value\">\n" + //
+				"  <!ELEMENT b|bb (#PCDATA)>\n" + //
+				"]>\n" + //
+				"<aaa>\n" + //
+				"  <bbb>Text Content</bbb>\n" + //
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(3, 2, 3, 26), r(0, 15, 4, 0), r(0, 0, 4, 2), r(0, 0, 8, 0)));
+	}
+
+	@Test
+	public void testDoctypeDeclaration8() {
+		// Only difference in this test with one above is the prolog
+		String xml = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n" + //
+				"<!DOCTYPE aaa [\n" + //
+				"  <!ELEMENT aaa (bbb)+>\n" + //
+				"  <!ATTLIST aaa key CDATA \"value\">\n" + //
+				"  <!ELEMENT b|bb (#PCDATA)>\n" + //
+				"]>\n" + //
+				"<aaa>\n" + //
+				"  <bbb>Text Content</bbb>\n" + //
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(4, 2, 4, 26), r(0, 15, 4, 0), r(0, 0, 4, 2), r(0, 0, 8, 0)));
+	}
+
+	@Test
+	public void testEmptyDoctypeDeclaration() {
+		String xml = "<!DOCTYPE aaa [\n" + //
+				"|]>\n" + //
+				"<aaa>\n" + //
+				"  <bbb>Text Content</bbb>\n" + //
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(0, 15, 1, 0), r(0, 0, 1, 2), r(0, 0, 5, 0)));
+	}
+
+	@Test
+	public void testEmptyDoctypeDeclaration2() {
+		String xml = "<!DOCTYPE aaa [|\n" + //
+				"]>\n" + //
+				"<aaa>\n" + //
+				"  <bbb>Text Content</bbb>\n" + //
+				"</aaa>\n";
+		testSelectionRange(xml, //
+				sr(r(0, 15, 1, 0), r(0, 0, 1, 2), r(0, 0, 5, 0)));
+	}
+
+	//#endregion
+
+	//#region test utils
+
+	private static final CancelChecker NULL_CHECKER = () -> {};
+
+	private void testSelectionRange(String xml, SelectionRange... selectionRanges) {
```

Comment:
Please move testSelectionRange in XMLAssert to be consistent with other features.